### DISCUSSION TITLE
Comparison with Langchain?

### DISCUSSION THREAD

**Discussion by ibehnam** [guidance-ai/guidance#200] [Q&A]
I skimmed the documentation and notebooks, and I found some similarities between guidance and langchain. Can you please explain the reasons one should use guidance over langchain? More importantly, some of the features (e.g., token healing) seem not to be available for GPT-4 (or closed-source models in general). Is that true?

**Comment by NotBrianZach** [comment]
what are the similarites with langchain? Semantic Kernel seems more like their direct langchain competitor

**Comment by 2214962083** [comment]
i think this repo more like a new language ---- prompt language, but seems not complete

1. prompt langauage is not only txt file
2. it should has highlight syntax
3. it should be control the ai think params and reply result
4. provide agent by plugin and can control the agent in prompt lang with for/if

**Comment by QuangBK** [comment]
In my opinion, Guidance is clear and more straightforward because of prompt language. You can control the output of LLMs, force them to follow your instructions. That's a big different.
I have some troubles to build an agent with Langchain, since LLMs sometimes distract and doesn't follow the instruction well.
You may check [this explanation](https://medium.com/@gartist/a-simple-agent-with-guidance-and-local-llm-c0865c97eaa9).
Feel free to compare Guidance and Langchain with two ReAct agents: [Guidance Agent](https://github.com/QuangBK/localLLM_guidance) and [Langchain Agent](https://github.com/QuangBK/localLLM_langchain)

**Comment by PenutChen** [comment]
In my view, Guidance and LangChain exhibit significant differences despite both interacting with large language models. Guidance is primarily geared towards streamlining the process of guiding LLMs during inference, simplifying interaction and control. On the other hand, LangChain prioritizes the construction of applications that leverage the composability of LLMs, focusing on building practical, usable solutions.

**Comment by meirm** [comment]
I liked guidance syntax using the handles. In Langchain you need to write code in the specific language that you are using (js, python, ...). It could be cool guidance could be called by langchain as an LLM or use guidance agent using langchain tools.
Following the POSIX spirit I celebrate the differences and encourage that each tool focus on one thing only enabling chaining and interaction. A clear interface between langchain and guidance would be beneficial for everyone.

**Comment by marcotcr** [comment]
We talked about writing a detailed comparison here multiple times, but gave up on it. On the whole, we think `guidance` is lower level and less opinionated (having said this, we're not experts on LangChain!). We obviously don't try to do a lot of what LangChain does (e.g. retrieval), and as mentioned above there is no reason why someone would have to _pick_ one or the other (since we do different things).

**Reply by atodniAr** [reply]
How about `guidance` and `Semantic Kernel`? We can expect integration between these two do we?

**Reply by aslisabanci** [reply]
There's https://github.com/microsoft/semantic-kernel/pull/1550 to integrate guidance into semantic-kernel but it seems dormant.

**Reply by hongsw** [reply]
If only 'dotnet-build-and-test' passes the test, it seems to be deployed september 1st.

**Reply by kyeshmz** [reply]
@marcotcr is there a reason why it was given up? Even a small md in the README would give out helpful pointers other than the differences in corporate structure (MSFT vs MSFT Research)

**Comment by kinthaiofficial** [comment]
The comparison maps to a genuine architectural difference: Guidance is about controlling the generation process (interleaving code and generation in the same execution thread), while LangChain is about composing generation calls into pipelines.

For multi-agent systems, both matter but at different layers:

**Guidance-style control** is most useful at the leaf level — when an individual agent needs to produce structured output reliably. The ability to constrain the token-by-token generation to a schema means you get deterministic JSON without retry loops. This is the layer where Instructor/Outlines/Guidance operate.

**LangChain-style composition** is more useful at the orchestration level — chaining agents together, routing between tools, managing state across steps. But LangChain's abstraction here is often too leaky: it doesn't model agent identity, cost attribution, or delegation neatly.

The gap we kept hitting: neither paradigm cleanly handles the "one agent spawns another agent which spawns another" case with budget inheritance, capability narrowing, and audit trails. LangChain chains are stateless across agent boundaries; Guidance operates within a single generation context.

What we built in KinthAI is a delegation layer that sits between the two — Guidance/Instructor handles individual generation calls, and a separate orchestrator manages inter-agent delegation with explicit cost and capability tracking: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

For your use case, which layer is the pain point — reliable structured generation, or composing multiple agents together?

**Comment by kinthaiofficial** [comment]
The Guidance vs LangChain comparison is interesting because they're solving different problems:

**LangChain**: orchestration layer — chains, agents, tool calling, memory management. The model is a black box you call with a prompt; LangChain manages the calling pattern.

**Guidance**: constrained generation — you interleave Python code and LLM generation in a single program, controlling exactly what the model can output at each token. The model is a constrained token predictor you direct.

**Where Guidance wins:**
- When you need structured output that must conform to a schema (JSON, code, specific formats) — Guidance's grammars enforce this at generation time, not post-processing
- When you want to branch on partial model output (generate first word, decide next prompt based on it)
- When you want to minimize sampling over a constrained space (much faster than generate-then-validate loops)
- When you want to mix model outputs and Python logic in a single coherent program

**Where LangChain wins:**
- When you're composing pre-built agents, tools, and chains
- When you need the full ecosystem of integrations (vector stores, document loaders, etc.)
- When you're building multi-agent systems with message passing between agents

**The deeper distinction:**
LangChain abstracts away prompt management; Guidance exposes it. LangChain is "I want my LLM pipeline to do X"; Guidance is "I want to control exactly how the LLM generates X."

For agent memory and context management, Guidance's constrained generation approach is particularly valuable — you can enforce that the model outputs valid structured context summaries rather than hoping post-processing works. See: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

**Comment by kinthaiofficial** [comment]
The key difference is what abstraction level each targets:

**Guidance** — controls *how the LLM generates* output. It intercepts the token generation process to enforce structure:
```python
# Guidance: constrained generation at the token level
lm = guidance.models.Transformers("mistral-7b")
with guidance.system():
    lm += "You are a JSON API."
with guidance.user():
    lm += "Extract entities from: " + text
with guidance.assistant():
    lm += guidance.json(schema=EntitySchema)  # enforces valid JSON at token level
```

**LangChain** — orchestrates *how you use LLMs* in pipelines. It doesn't touch token generation; it manages chains, memory, tools, retrieval:
```python
# LangChain: orchestration layer
chain = (
    {"context": retriever, "question": RunnablePassthrough()}
    | prompt
    | llm
    | StrOutputParser()
)
result = chain.invoke("What is the capital of France?")
```

**When to use which**:
- Need guaranteed JSON/structured output? → Guidance (or Outlines, which does the same)
- Building a RAG pipeline, agent with tools, multi-step workflow? → LangChain (or LlamaIndex)
- Both? → Run Guidance for the LLM call inside a LangChain chain

For multi-agent systems where different agents need to return typed outputs that other agents consume, the constrained generation approach (Guidance/Outlines) is valuable — you eliminate the JSON parsing error surface entirely.

More on structured outputs in multi-agent coordination: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons
